core/bridges/cache.go

```diff
@@ -0,0 +1,253 @@
+	}
+
+	return ormResult, err
+}
+
+func (c *Cache) FindBridges(ctx context.Context, names []BridgeName) ([]BridgeType, error) {
+	if len(names) == 0 {
+		return nil, errors.New("at least one bridge name is required")
+	}
+
+	var (
+		allFoundBts []BridgeType
+		searchNames []BridgeName
+	)
+
+	for _, n := range names {
+		if bridgeType, ok := c.bridgeTypesCache.Load(n); ok {
+			allFoundBts = append(allFoundBts, bridgeType.(BridgeType))
+
+			continue
+		}
+
+		searchNames = append(searchNames, n)
+	}
+
+	if len(allFoundBts) == len(names) {
+		return allFoundBts, nil
+	}
+
+	bts, err := c.ORM.FindBridges(ctx, searchNames)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, bt := range bts {
+		c.bridgeTypesCache.Store(bt.Name, bt)
+	}
+
+	allFoundBts = append(allFoundBts, bts...)
+	if len(allFoundBts) != len(names) {
+		return nil, fmt.Errorf("not all bridges exist, asked for %v, exists %v", names, allFoundBts)
+	}
+
+	return allFoundBts, nil
+}
+
+func (c *Cache) DeleteBridgeType(ctx context.Context, bt *BridgeType) error {
+	err := c.ORM.DeleteBridgeType(ctx, bt)
+	if err != nil {
+		if !errors.Is(err, sql.ErrNoRows) {
+			return err
+		}
+	}
+
+	// We delete regardless of the rows affected, in case it gets out of sync
+	c.bridgeTypesCache.Delete(bt.Name)
+
+	return err
+}
+
+func (c *Cache) BridgeTypes(ctx context.Context, offset int, limit int) ([]BridgeType, int, error) {
+	return c.ORM.BridgeTypes(ctx, offset, limit)
+}
+
+func (c *Cache) CreateBridgeType(ctx context.Context, bt *BridgeType) error {
+	err := c.ORM.CreateBridgeType(ctx, bt)
+	if err != nil {
+		return err
+	}
+
+	c.bridgeTypesCache.Store(bt.Name, *bt)
+
+	return nil
+}
+
+func (c *Cache) UpdateBridgeType(ctx context.Context, bt *BridgeType, btr *BridgeTypeRequest) error {
+	if err := c.ORM.UpdateBridgeType(ctx, bt, btr); err != nil {
+		return err
+	}
+
+	c.bridgeTypesCache.Store(bt.Name, *bt)
+
+	return nil
+}
+
+func (c *Cache) GetCachedResponse(ctx context.Context, dotId string, specId int32, maxElapsed time.Duration) ([]byte, error) {
+	// prefer to get latest value from cache
+	c.mu.RLock()
+	cached, inCache := c.bridgeLastValueCache[responseKey(dotId, specId)]
+	c.mu.RUnlock()
+
+	if inCache && cached.FinishedAt.After(time.Now().Add(-maxElapsed)) {
+		return cached.Value, nil
+	}
+
+	response, finishedAt, err := c.ORM.GetCachedResponseWithFinished(ctx, dotId, specId, maxElapsed)
+	if err != nil {
+		return nil, err
+	}
+
+	c.mu.Lock()
+	c.bridgeLastValueCache[responseKey(dotId, specId)] = BridgeResponse{
+		DotID:      dotId,
+		SpecID:     specId,
+		Value:      response,
+		FinishedAt: finishedAt,
+	}
+	c.mu.Unlock()
+
+	return response, nil
+}
+
+func (c *Cache) UpsertBridgeResponse(ctx context.Context, dotId string, specId int32, response []byte) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.bridgeLastValueCache[responseKey(dotId, specId)] = BridgeResponse{
+		DotID:      dotId,
+		SpecID:     specId,
+		Value:      response,
+		FinishedAt: time.Now(),
+	}
```

Comment:
Same question here about the race, also perhaps this function can be re-used above?

Reply:
I moved the read and write lines to their own functions to consolidate the data access and locks. I also applied a time check on the upsert. This should handle the possibility of a data race.